### PR TITLE
Migrate gemonlinedb away from Xerces and enter the JSON era

### PR DESCRIPTION
**Pull request by lmoureaux**
## Description

This PR is a nearly complete rewrite of the gemonlinedb serialization mechanism to migrate to a JSON-based system. The previous implementation based on Xerces had proven crash-prone and generally hard to maintain. The new JSON-based code is shorter and more readable, as illustrated by the line count of this PR: `965 insertions(+), 2174 deletions(-)` (code only).

The `dbloader` supports JSON input and the generated files are as close to what the `dbloader` expects as the previous XML (meaning that it takes some manual adjustments to make it work).

The JSON library used, [nlohmann/json](https://github.com/nlohmann/json), is installed system-wide (twice!) on `gem904daq02` and shipped with XDAQ 15. Hopefully the various versions won't clash.

JSON serialization will come handy when building the Web interface.

In the process of supporting JSON, more features have been added to the code generator in `parseDef.py`. As a consequence, the public API of several serialization objects has changed. I think it's fine because no merged code is using it.

Some previously working things have been broken in the process:

* `TStore` serialization is broken for the VFAT channels.
* XSD schemas have been dropped. If really needed, I can make [JSON schemas](https://json-schema.org/) later, but for now no validator is installed on the DAQ machines.
* The `KindOfPart` attribute of the `Part` object in the JSON has been lost. A small redesign is required to add it back.
* `std::exception` may leak out of `throw(...)` methods inherited from XDAQ, causing the program to abort. Since the `throw` clauses are gone in XDAQ 15, I hope they will be handled gracefully in the future.

Since this PR is huge, I'm not fixing these rather minor issues that require architectural changes.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context

See description.

## How Has This Been Tested?

Adding tests, running existing ones and running the Manager.

### Screenshots (if appropriate):

The Manager happily running on JSON:
image

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: gemonlinedb/src/common/AMC13Configuration.cc
Comment (jsturdy) [inline gemonlinedb/src/common/AMC13Configuration.cc]: Is this still needed?

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (jsturdy) [inline gemonlinedb/src/common/FileConfigurationProvider.cc]: likewise

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (jsturdy) [inline gemonlinedb/src/common/FileConfigurationProvider.cc]: likewise

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (jsturdy) [inline gemonlinedb/src/common/FileConfigurationProvider.cc]: likewise

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (jsturdy) [inline gemonlinedb/src/common/FileConfigurationProvider.cc]: can this `boost` stuff be replaced with a simple `try/catch`?
or will `loadInternal` fail differently if the file doesn't exist/fails to get parsed?

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (jsturdy) [inline gemonlinedb/src/common/FileConfigurationProvider.cc]: likewise

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (jsturdy) [inline gemonlinedb/src/common/FileConfigurationProvider.cc]: likewise

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (jsturdy) [inline gemonlinedb/src/common/FileConfigurationProvider.cc]: likewise

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (lmoureaux) [reply gemonlinedb/src/common/FileConfigurationProvider.cc]: I figured that checking the extension could be a useful semantic check (to check that the user isn't trying to load XML data or a GBT blob file...).
When `loadInternal` fails it throws `xcept::Exception` or `std::exception` derived classes.

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (lmoureaux) [reply gemonlinedb/src/common/FileConfigurationProvider.cc]: Fixed

File: gemonlinedb/src/common/AMC13Configuration.cc
Comment (lmoureaux) [reply gemonlinedb/src/common/AMC13Configuration.cc]: Fixed

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (jsturdy) [reply gemonlinedb/src/common/FileConfigurationProvider.cc]: right, but i could easily just have my blob/xml named `myxmlfile.json` (for unknown/nefarious reasons) and the check will pass, though it *should* still fail in the `loadInternal` call.
likewise, my valid `JSON` could be named `myjson.txt`, but this would fail.
so i was just wondering:
1) if the `boost` dependency was truly necessary (this was where i started, even though i know there are many other `boost` related things scattered throughout the code)
1) if the code could be simplified and rely on the `loadInternal` to handle invalid input data

**Comment by jsturdy** [comment]
Question about how much `xerces` is expected to remain in `gemonlinedb`:
```
egrep -i xerces -r gemonlinedb|wc    git:feature/json-migration(to↓with)
    133     626   13604
```
```
egrep -i xerces -r gemonlinedb|awk '{split($$0,a,":"); print a[1] }'|uniq                                                                                                                                                                                                                git:feature/json-migration(to↓with)
gemonlinedb/Makefile
gemonlinedb/include/gem/onlinedb/AMC13Configuration.h
gemonlinedb/include/gem/onlinedb/SystemTopology.h
gemonlinedb/include/gem/onlinedb/detail/XMLUtils.h
gemonlinedb/include/gem/onlinedb/gemonlinedb.h
gemonlinedb/src/common/AMC13Configuration.cc
gemonlinedb/src/common/ConfigurationManager.cc
gemonlinedb/src/common/SystemTopology.cc
gemonlinedb/src/common/XMLUtils.cc
gemonlinedb/test/testConfigurationLinker.cc
gemonlinedb/test/testConfigurationManager.cc
gemonlinedb/test/testSystemTopology.cc
```

File: gemonlinedb/src/common/FileConfigurationProvider.cc
Comment (lmoureaux) [reply gemonlinedb/src/common/FileConfigurationProvider.cc]: Code simplified and `boost` includes dropped.

**Review by lmoureaux — COMMENTED** [review]
> Question about how much `xerces` is expected to remain in `gemonlinedb`:

This PR doesn't address the topology file (still using XML), so a bit of Xerces is expected to remain.

**Review by jsturdy — APPROVED** [review]
OK, I'll leave it to you to track what `xerces-c` dependencies remain and how/if you want to phase them out.
As it is, I think it will make sense to merge this and then rebase #311 on top before merging that

**Pull request merged**
